### PR TITLE
Add TestPyPI release workflow

### PR DESCRIPTION
**Pull request by acisseJZhong**
## Summary

  Add a manually triggered workflow to stage release candidates on TestPyPI before production release.

  The workflow:

  - verifies the version is an RC
  - builds and checks the wheel and sdist
  - publishes through the protected `test-release` environment using OIDC
  - validates installation on Python 3.11 and 3.12
  - runs a short CPU training smoke test and checks loss decreases

  ## Usage

  1. Open **Actions -> Publish a Release to TestPyPI**.
  2. Select the release branch, for example `release/0.3`.
  3. Click **Run workflow**.
  4. Approve the `test-release` deployment.
  5. Confirm publishing and RC validation succeed before creating the production GitHub Release.

  ## Test plan

  - Validated the workflow YAML.
  - Ran `git diff --check`.
  - Verified the CPU training smoke test produces finite, decreasing loss.

### CODE REVIEW

File: .github/workflows/test_release.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# Manual TestPyPI dry-run, mirrors release.yml but targets TestPyPI (staging) instead of PyPI.
+# Run this via "Run workflow" BEFORE cutting the real GitHub Release, verify the upload installs
+# and renders on test.pypi.org, then cut the release (which triggers release.yml -> real PyPI).
+
+name: Publish a Release to TestPyPI
+
+on:
+  workflow_dispatch:
```
Comment (tianyu-l) [inline .github/workflows/test_release.yml:8]: How is this going to detect v0.3? Do we first need to create a new release at https://github.com/pytorch/torchtitan/releases/new?

Could you share more on the overall workflow?

Comment (acisseJZhong) [reply .github/workflows/test_release.yml:8]: the test_release workflow will push the torchtitan 0.3.0rc1 to test pypi(because when I trigger this workflow, I will select the release/0.3 branch), after the package got pushed to test pypi, it will trigger the validate_rc workflow to do some basic testing, e.g. import check, check model can run and loss decrease on cpu. 

After this PR got merged, I can trigger the workflow and test out the process. Once it works, I will update the release readme. It's possible that this needs iteration, but I cannot test without merging this PR first.

**Review by tianyu-l — APPROVED** [review]
